### ISSUE TITLE
Tool Outputs in Attention Mask for Log Probabilities

### ISSUE THREAD

**Issue by michalsr** [Time-Search/TimeSearch-R#7]
From my understanding, the attention mask that is passed into https://github.com/Time-Search/TimeSearch-R/blob/main/time_r1/trainer/grpo_trainer_env.py#L874 `(def _get_per_token_logps(self, model, input_ids, attention_mask, multimodal_inputs, logits_to_keep, batch_size=None) -> torch.Tensor:`
is defined on line 1163 
 `completion_mask = compute_tool_response_mask(completion_ids) * completion_mask` which masks out tool outputs. 

Why are the tool outputs not included in the attention mask for computing the log probabilities? Shouldn't they just be masked out for the loss?

**Comment by DesertsP** [comment]
The tool-token mask should only be applied to the loss (i.e., which tokens contribute to the GRPO objective). Masking those tokens in the attention_mask (the mask used to compute logits / per-token logprobs) makes _get_per_token_logps return log-probs computed without the tool outputs as context, so the per-token log-probs are incorrect for any computation that expects the model to have seen those tool tokens.

In this repo's current code we assert that num_iterations == 1 (and steps_per_generation == 1). Under that assumption, the only place the result of _get_per_token_logps is used is to compute per_token_kl (the KL term). That is why the bug has a limited immediate effect, but it is still incorrect and should be fixed.

NOTE: coef_1 = torch.exp(per_token_logps - old_per_token_logps) is always ones.
